### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by wooorm**
-   Refactor list of authors to additionally include
  other mdast authors,  and re-order.
  
  More info:
  
  ```
  https://github.com/tj/git-extras/blob/
  c8aca55cd70b52079fd76703ba3b54aaef8fe9fe/bin/git-authors
  ```
-   Change copyright of `watch-output-cache` to me. Although I really
  appreciate your work, and do recognize it (e.g., in AUTHORS) and
  in the repo, if you assign these rights over it makes future changes
  without consulting you easier. Although I phrase this as a question,
  I do not think having code copyrighted to others is viable. Thus,
  for me to merge this, this change is of vital importance.
  
  More info:
  
  ```
  http://haacked.com/archive/2006/01/26/
  WhoOwnstheCopyrightforAnOpenSourceProject.aspx/
  ```
-   I refactor the cache to be a class, this allows multiple watching
  mdast instances in a single process.
-   I refactored the spinner to spin when waiting between `run`s,
  I found that, when actually running, the thread was so full that
  the spinner got stuck. Now, when watching, you see the spinner
  spinning which I really like (try it out, this is IMO for the
  better!)
  
  In addition, the spinner is extracted into a new file, which
  contains all code needed for it to work.
-   I moved functions which were hoisted above the place they
  were used.
-   I moved variable declarations to the start of their scope.
-   I refactored some comments to be less verbose. I appreciate
  your extensiveness in these comments, it helps me understand why
  things happen this way, but they were very verbose.
  
  Some were moved to the place where entities were created,
  e.g., `sourcePaths`, and others were rewritten.
-   I refactor the warning message when files are cached instead of
  written. This change ensures it’s only shown when needed.
  
  I additionally split it into two: a message about quitting
  the watch using `CTRL+C` (which is always shown), and the warning
  when applicable.
-   Docs.

### CODE REVIEW

File: AUTHORS [hunk 1]
```diff
@@ -7,11 +8,13 @@ Kitson Kelly <dojo@kitsonkelly.com>
 Hsiaoming Yang <me@lepture.com>
 MinRK <benjaminrk@gmail.com>
 Isaac Z. Schlueter <i@izs.me>
+Ben Briggs <beneb.info@gmail.com>
 Ian Storm Taylor <ian@ianstormtaylor.com>
 Brad Harris <bmharris@gmail.com>
 Maksim Kozhukh <mkozhukh@ya.ru>
 Martin Zagora <zaggino@gmail.com>
 Matias Niemelä <matias@yearofmoo.com>
+Eugene Sharygin <eush77@gmail.com>
```
Comment (why-jay) [inline AUTHORS:17]: I'm totally cool with adding authors, but did Ben and Eugene contribute to this specific PR? Shouldn't they be added in via a separate commit?

File: doc/mdast.1.md [hunk 2]
```diff
@@ -97,7 +97,7 @@ COMMAND LINE USAGE for more information.
 ### `-e`, `--ext` <_extensions_>
 
 ```sh
-mdast --ext .doc
+mdast --ext doc
```
Comment (why-jay) [inline doc/mdast.1.md:100]: Is this line's change relevant to this PR?

Comment (why-jay) [inline doc/mdast.1.md:100]: (by "this PR" I mean the watch PR as a whole)

File: AUTHORS (hunk 1 above)
Comment (wooorm) [inline AUTHORS:17]: Yeah, you can ignore those, I can either change this PR or you can remove them before squashing  :+1:

File: doc/mdast.1.md (hunk 2 above)
Comment (wooorm) [inline doc/mdast.1.md:100]: Nope, same as the authors one!

File: lib/cli/watch-output-cache.js [hunk 3]
```diff
@@ -1,46 +1,73 @@
 /**
  * @author YJ Yang
- * @copyright 2015 YJ Yang
+ * @copyright 2015 Titus Wormer
```
Comment (why-jay) [inline lib/cli/watch-output-cache.js:3]: No worries about this one, I'm cool with it

File: lib/cli/watch-output-cache.js [hunk 4]
```diff
@@ -1,46 +1,73 @@
...
 
 /**
  * Add a VFile to the cache so that its content can later be used.
  *
+ * @this {Cache}
  * @param {VFile} file - The file to add to the cache.
  */
 function add(file) {
-    var filePath = file.filePath();
+    var filePath = file.sourcePath || file.filePath();
```
Comment (why-jay) [inline lib/cli/watch-output-cache.js:51]: Can you tell me why using `sourcePath` in addition is important here?

File: lib/cli/watch-output-cache.js (hunk 3 above)
Comment (wooorm) [inline lib/cli/watch-output-cache.js:3]: :+1:

File: lib/cli/file-set-pipeline/log.js [hunk 5]
```diff
@@ -12,52 +12,43 @@
...
 var report = require('vfile-reporter');
 
 /**
+ * Check whether `file` is provided by the user.
+ *
+ * @param {File} file - Virtual file.
+ * @return {boolean?} - whether `file` is provided by the
+ *   user.
+ */
+function isProvidedByUser(file) {
```
Comment (why-jay) [inline lib/cli/file-set-pipeline/log.js:25]: Don't wanna nitpick, but is extracting this function out really necessary? It's a tiny function that was doing alright as an anonymous function. And this change is not really relevant to the watch PR. It will only confuse our commit diff at the end, in my opinion

File: lib/cli/watch-output-cache.js (hunk 4 above)
Comment (wooorm) [inline lib/cli/watch-output-cache.js:51]: Hmm, sucks that there’s no coverage for the CLI yet.
Anyway, I’m thinking now that this isn’t necessary: `sourcePath` isn’t available because the files haven’t been added to the `fileSet` yet...

File: lib/cli/file-set-pipeline/log.js (hunk 5 above)
Comment (wooorm) [inline lib/cli/file-set-pipeline/log.js:25]: Okay ignoring it is fine, I added it hear because it makes the code look nicer (the filter can be on one line).

Also: great that you’re nitpicking ;)

File: lib/cli/index.js [hunk 6]
```diff
@@ -16,148 +16,105 @@ var chalk = require('chalk');
...
-         * In contrast, if the following is called also on the *first*
-         * pipeline run, then you will see an extra spinner.
+         * Watch files source-locations of files in
+         * the file-set.
          */
-        cli.stdout(getSpinnerFrame());
-    }
+
+        watcher = chokidar.watch(cli.fileSet.sourcePaths, {
+            'ignoreInitial': true
```
Comment (why-jay) [inline lib/cli/index.js:103]: The quotes around `ignoreInitial` aren't necessary?

Comment (wooorm) [inline lib/cli/index.js:103]: That’s style, IMO. I always quote props, just so I never have to think about it.

File: lib/cli/watch-output-cache.js (hunk 4 above)
Comment (why-jay) [inline lib/cli/watch-output-cache.js:51]: OK, after I merge your PR, I'll test, and will remove the `sourcePath` part if that seems like the right thing to do

File: lib/cli/index.js (hunk 6 above)
Comment (why-jay) [inline lib/cli/index.js:103]: Oh OK, interesting. I didn't know that all this time reading your codebase haha

Comment (wooorm) [inline lib/cli/index.js:103]: Oh weird, I think I do it everywhere though. Maybe not that many object literals in mdast? 😉
